primitives/blockchain/src/error.rs

```diff
@@ -35,110 +35,150 @@ pub enum ApplyExtrinsicFailed {
 	NotAvailableOnLightClient,
+
 	/// Invalid remote CHT-based proof.
 	#[error("Remote node has responded with invalid header proof")]
 	InvalidCHTProof,
+
 	/// Remote fetch has been cancelled.
 	#[error("Remote data fetch has been cancelled")]
 	RemoteFetchCancelled,
+
 	/// Remote fetch has been failed.
 	#[error("Remote data fetch has been failed")]
 	RemoteFetchFailed,
+
 	/// Error decoding call result.
 	#[error("Error decoding call result of {0}")]
 	CallResultDecode(&'static str, #[source] CodecError),
+
 	/// Error converting a parameter between runtime and node.
 	#[error("Error converting `{0}` between runtime and node")]
 	RuntimeParamConversion(String),
+
 	/// Changes tries are not supported.
 	#[error("Changes tries are not supported by the runtime")]
 	ChangesTriesNotSupported,
+
 	/// Error reading changes tries configuration.
 	#[error("Error reading changes tries configuration")]
 	ErrorReadingChangesTriesConfig,
+
 	/// Key changes query has failed.
 	#[error("Failed to check changes proof: {0}")]
 	ChangesTrieAccessFailed(String),
+
 	/// Last finalized block not parent of current.
 	#[error("Did not finalize blocks in sequential order.")]
 	NonSequentialFinalization(String),
+
 	/// Safety violation: new best block not descendent of last finalized.
 	#[error("Potential long-range attack: block not in finalized chain.")]
 	NotInFinalizedChain,
+
 	/// Hash that is required for building CHT is missing.
 	#[error("Failed to get hash of block for building CHT")]
 	MissingHashRequiredForCHT,
+
 	/// Invalid calculated state root on block import.
 	#[error("Calculated state root does not match.")]
 	InvalidStateRoot,
+
 	/// Incomplete block import pipeline.
 	#[error("Incomplete block import pipeline.")]
 	IncompletePipeline,
+
+	/// Transaction pool initizliation is not complete.
 	#[error("Transaction pool not ready for block production.")]
 	TransactionPoolNotReady,
+
+	/// Database yielded an error.
 	#[error("Database")]
 	DatabaseError(#[from] sp_database::error::DatabaseError),
-	/// A convenience variant for String
-	#[error("{0}")]
-	Msg(String),
-}
 
-impl<'a> From<&'a str> for Error {
-	fn from(s: &'a str) -> Self {
-		Error::Msg(s.into())
-	}
-}
+	#[error("Failed to get header for hash {0}")]
+	MissingHashInHeader(String),
 
-impl From<String> for Error {
-	fn from(s: String) -> Self {
-		Error::Msg(s)
-	}
+	#[error("Failed to load the block weight for block {0}")]
+	MissingBlockWeightInHeader(String),
+
+	#[error("WASM override IO error")]
+	WasmOverrideIo(PathBuf, #[source] std::io::Error),
+
+	#[error("Overwriting WASM requires a directory where local \
+	WASM is stored. {} is not a directory", .0.display())]
+	WasmOverrideNotADirectory(PathBuf),
+
+	#[error("Duplicate WASM Runtimes found: \n{}\n", .0.join("\n") )]
+	DuplicateWasmRuntime(Vec<String>),
+
+	#[error(transparent)]
+	Foreign(#[from] Box<dyn std::error::Error + Send + Sync + 'static>),
+
+	/// Should be avoided if possible, use `Foreign` instead.
```

Comment:
I'd rather use the `deprecated` attribute here or (a more radical measure) remove it.
Is this message addressed to devs of substrate or someone who's building on top?

Reply:
From the first review from basti, the way to go was to not tag it with `#[deprecated(note = "cause it's bad")` but to remove it entirely. I removed the stringified versions where possible and adjusted the doc comments otherwise.

Reply:
Where do we still need this?

Reply:
Same as above, it's the `.into()`s

Reply:
And when will this be removed?

Reply:
Planning to do the chore on another Friday afternoon

Reply:
Implemented typed errors down to a certain level